source/core/Synchronization.cpp

```diff
@@ -452,123 +452,174 @@ VIREO_FUNCTION_SIGNATURE1(SetOccurrence, OccurrenceRef)
-
-    if (space >= additionalCount) {
-        // There is enough room, wrap the insert location as needed.
-        if (insert == RemoveIndex() && _count > 0) {
-            // Insert at opposite end (front), shift data to make room
-            AQBlock1 *pData = _elements->BeginAt(insert);
-            TypeRef elemType = _elements->ElementType();
-            elemType->CopyData(pData, (AQBlock1*)pData + additionalCount * elemType->TopAQSize(), _count);
-        } else if (_insert >= length) {
-            _insert = 0;
-        }
-        return true;
-    } else {
-        // Not enough room, grow (if possible)
-        if (_maxSize > 0 && length + additionalCount > _maxSize)
+    TypeRef eltType = _elements->ElementType();
+    VIREO_ASSERT(position < _elements->Length());
+    void* pTarget = _elements->BeginAt(position);
+    eltType->CopyData(pData, pTarget);
+    _count++;
+    VIREO_ASSERT(_count <= _elements->Length());
+    VIREO_ASSERT(_maxSize < 0 || _count <= _maxSize);
+}
+
+Boolean QueueCore::ResizeInternalBufferIfEmpty() const
+{
+    VIREO_ASSERT(_front == -1 && _back == -1 && _count == 0);
+    if (_elements->Length() == 0) {  // _maxSize must be at least 1 or negative, so grow
+        NIError err = _elements->Insert1D(0, 1);
+        if (err != kNIError_Success) {
             return false;
-        NIError err = _elements->Insert1D(insert, additionalCount);
-        return (err == kNIError_Success);
+        }
     }
+    return true;
 }
+
 //------------------------------------------------------------
+// Insert at the back of queue
 Boolean QueueCore::Enqueue(void* pData)
 {
-    /* ???
-     This needs to be refactored.  If the Queue is not fixed/bounded, the array will continue to grow
-     in size to the maximum count of items ever enqueued, even if most or all of the items are dequeued
-     as they are queued so the conceptual size of the queue is always small.
-     (This was probably the idea behind the unimplemented Compress() method in the header.)  -CS */
-
-    if (!TryMakeRoom(1, _insert))
-        return false;
+    VIREO_ASSERT(_maxSize > 0 || _maxSize == -1);
+    if (_back == -1) {
+        if (!ResizeInternalBufferIfEmpty()) {
+            return false;
+        }
+        _back = _front = 0;
 
-    TypeRef eltType = _elements->ElementType();
-    void* pTarget = _elements->BeginAt(_insert);
-    eltType->CopyData(pData, pTarget);
+    } else {
+        IntIndex insert = (_back + 1) % _elements->Length();
+        if (insert != _front) {
+            _back = insert;
+        } else {  // Array full
+            VIREO_ASSERT(_count == _elements->Length());
+            if (_maxSize > 0 && _count == _maxSize)
+                return false;
+            if (insert == 0) {
+                // Just an optimization to grow the underlying array at the end than shifting all the elements by 1
+                insert = _elements->Length();
+            }
+            NIError err = _elements->Insert1D(insert, 1);
```

Comment:
Maybe consider adjusting the capacity here two grow 'smartly'?  Double the number of elements until you get to some large cap (1M?) and then become additive.  I just checked and noticed that the ResizeCapacity member of TypedArrayCore is private, though, so this might require more work and I think it's okay to put it off with a TO-DO and issue files.

Reply:
Good point. Created an issue to address this later.